vision_evaluation/evaluators.py

```diff
@@ -263,10 +268,79 @@ def __init__(self, prediction_filter):
     def _get_id(self):
         return f'precision_{self.prediction_filter.identifier}'
 
+    def calculate_score(self, average='macro'):
+        return super(PrecisionEvaluator, self).calculate_score(average=average, filter_out_zero_tgt=False)
+
     def _calculate(self, targets, predictions, average):
+        """
+        There is a special case that needs to be handled appropriately.
+        Due to filtering conditions, there are occasions where only 1 class remains in targets/predictions and sklearn interpretes this as an invalid configuration for multilabel.
+        Since when average == 'samples' is only calculated for multilabel, sm.recall_score throws an exception. e.g.:
+        predictions = np.array([[True]])
+        targets = np.array([[True]])
+        recall_score(targets, predictions, average='samples') throws the following error:
+        ValueError: Samplewise metrics are not available outside of multilabel classification.
+        """
+        if targets.shape[1] == 1 and average == 'samples':
+            targets = np.append(targets, np.zeros((targets.shape[0], 1)), axis=1)
+            predictions = np.append(predictions, np.zeros((predictions.shape[0], 1)), axis=1)
         return sm.precision_score(targets, predictions, average=average)
 
 
+class MeanAveragePrecisionAtK(MemorizingEverythingEvaluator):
```

Comment:
Does adding a prediction filter to existing AveragePrecisionEvaluator and calculate the AP@K with sample flavor work?

I tried the implementation mentioned above but it gives a different results than your tests, however, I cannot understand your implementation very well...

Reply:
Unfortunately no, it just does a different calculation. I experimented a bunch but it's just not the same. it makes sense, since from the doc: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html

AP summarizes a precision-recall curve as the weighted mean of precisions achieved at each threshold, with the increase in recall from the previous threshold used as the weight